### PR TITLE
Feature #156537808 - Add links to publications

### PR DESCRIPTION
**Pull request by monicajianu**
* Implemented EuropePMC client to retrieve publication information using Pubmed IDs or DOIs
* Added column for DOIs in experiment tables
* Implemented UI to display publications with associated titles, authors and links

### CODE REVIEW

**Comment by alfonsomunozpomer** [comment]
I can see two failing tests in gxa:
http://bar:8085/browse/ATLAS-GXACIDEVTASKS11-6/test

Can you have a look at them first, please?

**Comment by monicajianu** [comment]
All builds are now passing:

http://bar:8085/browse/ATLAS-GXACIDEVTASKS11
http://bar:8085/browse/ATLAS-SCXACIDEVTASKS11

File: base/src/test/java/uk/ac/ebi/atlas/utils/EuropePmcClientIT.java [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+    private EuropePmcClient subject;
+
+    @Test
+    public void publicationForValidDoi() {
+        Optional<Publication> result = subject.getPublicationByDoi("10.1126/sciimmunol.aan8664");
+
+        assertThat(result.isPresent(), is(true));
+        assertThat(result.get().getDoi(), is("10.1126/sciimmunol.aan8664"));
+        assertThat(result.get().getAuthors(), not(isEmptyString()));
+        assertThat(result.get().getTitle(), not(isEmptyString()));
```
Comment (alfonsomunozpomer) [inline base/src/test/java/uk/ac/ebi/atlas/utils/EuropePmcClientIT.java]: These are the sort of tests where I think AssertJ really shines, e.g.:
```
assertThat(result.orElseThrow(RuntimeException::new))
        .extracting("doi", "authors", "title")
        .contains("10.1126/sciimmunol.aan8664", "Patil VS, Madrigal A, Schmiedel BJ, Clarke J, O'Rourke P et al.", "Precursors of human CD4+ cytotoxic T lymphocytes identified by single-cell transcriptome analysis.");
```

File: base/src/main/java/uk/ac/ebi/atlas/model/Publication.java [hunk 2]
```diff
@@ -1,11 +1,24 @@
 package uk.ac.ebi.atlas.model;
 
+import org.codehaus.jackson.annotate.JsonIgnoreProperties;
+import org.codehaus.jackson.annotate.JsonProperty;
```
Comment (alfonsomunozpomer) [inline base/src/main/java/uk/ac/ebi/atlas/model/Publication.java]: I like Jackson! I can see ourselves gradually moving away from Gson in the future.

File: base/src/test/java/uk/ac/ebi/atlas/utils/EuropePmcClientIT.java [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+    }
+
+    @Test
+    public void noResultForEmptyPubmedId() {
+        Optional<Publication> result = subject.getPublicationByPubmedId("");
+
+        assertThat(result.isPresent(), is(false));
+    }
+
+}
```
Comment (alfonsomunozpomer) [inline base/src/test/java/uk/ac/ebi/atlas/utils/EuropePmcClientIT.java]: Can you add a test to get 100% coverage?

File: base/src/main/java/uk/ac/ebi/atlas/model/Publication.java [hunk 4]
```diff
@@ -51,18 +54,30 @@ public void setTitle(String title) {
...
+            }
+            else {
+                String shownAuthors = String.join(",", Arrays.copyOfRange(authorsArray, 0, 5));
 
-            return String.join(" ",shownAuthors, "et al.");
+                return String.join(" ",shownAuthors, "et al.");
+            }
         }
+
+        return authors;
```
Comment (alfonsomunozpomer) [inline base/src/main/java/uk/ac/ebi/atlas/model/Publication.java]: You’re returning `null` here. I don’t think this is what you want. 😉

File: sc/src/main/java/uk/ac/ebi/atlas/experimentpage/ExperimentController.java [hunk 5]
```diff
@@ -40,12 +40,19 @@ public String showExperimentPage(Model model,
         Experiment<?> experiment = experimentTrader.getExperiment(experimentAccession, accessKey);
         model.addAllAttributes(experiment.getAttributes());
 
+        if (!experiment.getDois().isEmpty()) {
+            model.addAttribute("publications", experimentPageContentService.getPublications(experiment.getDois()));
+        }
+        else if (!experiment.getPubMedIds().isEmpty()) {
+            model.addAttribute("publications", experimentPageContentService.getPublications(experiment.getPubMedIds()));
+        }
```
Comment (alfonsomunozpomer) [inline sc/src/main/java/uk/ac/ebi/atlas/experimentpage/ExperimentController.java]: Why isn’t the field `publication` populated as part of `experiment.getAttributes`?

File: sc/src/main/java/uk/ac/ebi/atlas/experimentpage/ExperimentPageContentService.java [hunk 6]
```diff
@@ -83,7 +91,16 @@ public JsonArray getAnalysisMethodsAsJson(String experimentAccession) {
         return result;
     }
 
-    private JsonObject getExperimentFileAsJson(ExperimentFileType experimentFileType, String experimentAccession, String accessKey) {
+    public List<Publication> getPublications(List<String> identifiers) {
```
Comment (alfonsomunozpomer) [inline sc/src/main/java/uk/ac/ebi/atlas/experimentpage/ExperimentPageContentService.java]: I find it a bit weird that all methods in `ExperimentPageContentService` return `JsonElement`s but this one. See, in relation to this, my comment above in `ExperimentController.java`. I think the design could be improved to be more consistent.

File: base/src/test/java/uk/ac/ebi/atlas/utils/EuropePmcClientIT.java [hunk 7]
```diff
@@ -25,27 +23,33 @@
...
-        assertThat(result.get().getDoi(), is("10.1126/sciimmunol.aan8664"));
-        assertThat(result.get().getAuthors(), not(isEmptyString()));
-        assertThat(result.get().getTitle(), not(isEmptyString()));
+        assertThat(result.isPresent()).isTrue();
+
+        assertThat(result.orElseThrow(RuntimeException::new))
+                .extracting("doi", "authors", "title")
+                .isNotEmpty();
+
+        assertThat(result.orElseThrow(RuntimeException::new).getDoi()).isEqualToIgnoringCase("10.1126/sciimmunol.aan8664");
```
Comment (alfonsomunozpomer) [inline base/src/test/java/uk/ac/ebi/atlas/utils/EuropePmcClientIT.java]: You can conflate the two tests:
```
        assertThat(result.orElseThrow(RuntimeException::new))
                .hasFieldOrPropertyWithValue("doi", "10.1126/sciimmunol.aan8664")
                .extracting("doi", "authors", "title")
                .isNotEmpty();
```

You don’t get the benefit of case insensitive match, though.

**Comment by alfonsomunozpomer** [comment]
Could you add tests for `ExperimentAttributesService`? You can take them away from `Experiment` and also remove `getAttributes()` so that all classes use the new service.

**Comment by alfonsomunozpomer** [comment]
This one spun a bit out of control, but the result is certainly better than what we have. Atlas developers from the future will feel grateful to your PR for certain!

**Pull request merged**